### ISSUE TITLE
Allow parameter only if another parameter is present.

### ISSUE THREAD

**Issue by dblock** [ruby-grape/grape#958]
For example:

```ruby
requires :product_id, type: String # a product id
optional :shelf_id, type: String # only if the product is on this shelf
optional :shelf_reachable, type: Boolean # only if the shelf is reachable
```

You shouldn't be able to specify `shelf_reachable` unless `:shelf_id` is present.

**Comment by u2** [comment]
That's a good idea.I just have this demand.
```
        params do
          requires :cid, type: Integer
          requires :car_order, type: Hash do
            optional :cars_params, type: Array do
              requires :id, type: Integer
              requires :quantity, type: Integer #
            end
            optional :car_sets_params, type: Array
          end
```
We want at least one of `cars_params` and `car_sets_params` is present.

**Comment by dblock** [comment]
@u2 We already have `at_least_one_of`.

**Comment by u2** [comment]
Thank you.:smile:

**Comment by rnubel** [comment]
After thinking about this for a bit, I think it'd be more useful to be able to say "if `shelf_id` is given, then require `shelf_reachable`" rather than the inverse. The two are closely related, at any rate, and I could imagine there being use cases for both. So along those lines, I've quickly prototyped a couple possible ways to implement such logic:

### Option 1 (DSL style)
https://github.com/rnubel/grape/compare/dependent_parameters_dsl
```ruby
params do
  optional :shelf_id, type: Integer
  if_given :shelf_id do
    requires :shelf_reachable, type: Boolean
  end
end
```

### Option 2 (Validation style)
https://github.com/rnubel/grape/compare/dependent_parameters_validation
```ruby
params do
  optional :shelf_id, type: Integer
  optional :shelf_reachable, type: Boolean
  if_given :shelf_id, require: :shelf_reachable
end
```

What do you think?

**Comment by dblock** [comment]
I like the first one, but I don't love the name. Maybe just `with`?

**Comment by rnubel** [comment]
Hmm... `with` is a bit vague. I think that's the tricky part here -- we want a term that's clear about what it means, but still concise. What about just `given`?

```ruby
params do
  optional :shelf_id, type: Integer
  given :shelf_id do
    requires :shelf_reachable, type: Boolean
  end
end
```

**Comment by dblock** [comment]
I like `given`, go for it!

**Comment by chrisyoumans** [comment]
Should `given` work within a nested parameter? E.g.:

```
requires :document, type: Hash do
    optional :file_contents, type: String
    given :file_contents do
        requires :file_name, type: String
    end
    optional :other_text, type: String
end
```

Looking at params_scope_spec, it seems like it's intended to. However I'm receiving "undefined method `given' for #<Grape::Validations::ParamsScope:0x007fa01124c198> (NoMethodError)."

**Comment by dblock** [comment]
This was resolved in #1047.
